pkg/network/common/dns.go

```diff
@@ -122,58 +133,110 @@ func (d *DNS) updateOne(dns string) (bool, error) {
 	return changed, nil
 }
 
-func (d *DNS) getIPsAndMinTTL(domain string) ([]net.IP, time.Duration, error) {
+func (d *DNS) doOneQuery(server, domain string, rtype uint16) ([]net.IP, int, error) {
 	ips := []net.IP{}
-	ttlSet := false
-	var ttlSeconds uint32
+	ttl := defaultTTL
 
-	for _, server := range d.nameservers {
-		msg := new(dns.Msg)
-		msg.SetQuestion(dns.Fqdn(domain), dns.TypeA)
+	msg := new(dns.Msg)
+	msg.SetQuestion(dns.Fqdn(domain), rtype)
 
-		dialServer := server
-		if _, _, err := net.SplitHostPort(server); err != nil {
-			dialServer = net.JoinHostPort(server, d.port)
-		}
-		c := new(dns.Client)
-		c.Timeout = 5 * time.Second
-		in, _, err := c.Exchange(msg, dialServer)
-		if err != nil {
-			return nil, defaultTTL, err
-		}
-		if in != nil && in.Rcode != dns.RcodeSuccess {
-			return nil, defaultTTL, fmt.Errorf("failed to get a valid answer: %v", in)
+	c := new(dns.Client)
+	c.Timeout = d.timeout
+	in, _, err := c.Exchange(msg, server)
+	if in == nil || err != nil {
+		return ips, ttl, err
+	}
+	if in.Rcode != dns.RcodeSuccess {
+		return ips, ttl, fmt.Errorf("failed to get a valid answer: %v", in)
+	}
+
+	for _, a := range in.Answer {
+		aTTL := int(a.Header().Ttl)
+		if aTTL < ttl && aTTL != 0 {
```

Comment:
This always returns the lowest TTL without considering if it's an A or AAAA record.

Reply:
hm... all of the responses in `in.Answer` _should_ be the same type (`rtype`)... I don't remember if the server is allowed to return answers of the wrong type? Easy enough to be defensive here though.

Reply:
Actually, the code should probably be considered correct as-is; in particular, if there is a CNAME and an A, and the CNAME has a lower TTL than the A, then we need to re-query when the CNAME expires, not when the A expires. So it's best to just take the lowest TTL of all returned answers.

Reply:
You're right.